helix-term/src/commands.rs

```diff
@@ -3282,7 +3282,17 @@ pub fn code_action(cx: &mut Context) {
                         lsp::CodeActionOrCommand::CodeAction(code_action) => {
                             log::debug!("code action: {:?}", code_action);
                             if let Some(ref workspace_edit) = code_action.edit {
-                                apply_workspace_edit(editor, offset_encoding, workspace_edit)
+                                log::debug!("edit: {:?}", workspace_edit);
+                                apply_workspace_edit(editor, offset_encoding, workspace_edit);
+                            }
+                            // if code action provides both edit and command first the edit
+                            // should be applied and then the command
+                            if let Some(command) = code_action.command {
+                                log::debug!("command: {:?}", command);
+
+                                tokio::spawn(async move {
+                                    language_server.command(command).await
+                                });
```

Comment:
The problem you're seeing is because you're trying to move `language_server` into the async closure.

This should work:

```suggestion
                                let future = language_server.command(command);
                                tokio::spawn(async move {
                                    future.await
                                    // TODO: log the result here 
                                });
```

Reply:
Tried that before, this moves the lifetime issue to

```
error[E0621]: explicit lifetime required in the type of `cx`
    --> helix-term/src/commands.rs:3268:30
     |
3268 |                 let picker = Picker::new(
     |                              ^^^^^^^^^^^ lifetime `'static` required
```

Reply:
Right, you're also reusing the language_server from outside the `cx` callback, and the compiler can't know how long the callback will live for, so `language_server` -> `doc` -> `cx` need to live for static.

You need to re-fetch the `doc`/`language_server` inside the callback, easiest way to do that is `let (doc, _view) = current!(editor)`, the most correct would be to `let doc_id = doc.id()` before `cx.callback`, then re-fetch that exact ID in the closure

Reply:
Thanks a lot for the help! Fixed now. I also extracted the command execution into separate function so it can be re-used so it can be re-used between commands, code actions, and possibly code lenses in the future.